### PR TITLE
fix : simple parser upgraded for table rendering

### PR DESCRIPTION
**Pull request by focaloid-dev**
**Related Issue :** https://oat-sa.atlassian.net/browse/AUT-560 
If user added the Table via CK Editor to the Passage during it's authoring, this Table will not display if the passage containing it will be added to the item during Item Authoring.

**Preconditions**: Passage containing table is created.

**Steps to reproduce:**

- Go to Items tab.
- 
- Create a new item and go to it's authoring.
- 
- Add any interaction to the canvas.
- 
- Add a Passage containing Table to this interaction.

**Actual result:** The Passage is added, the Table is not displayed in Item Authoring page.
**Expected result:** The Passage is added, all of it's contend is displayed including Table.

Env : http://tao-test-alshoja.playground.kitchen.it.taocloud.org:46031/

### CODE REVIEW

File: src/qtiItem/helper/simpleParser.js [hunk 1]
```diff
@@ -102,40 +103,55 @@ function buildTooltip(targetHtml, contentId, contentHtml) {
...
-    _.each(_parsableElements, function(qtiClass) {
-        $container.find(qtiClass).each(function() {
+    _.each(_parsableElements, function (qtiClass) {
+        $container.find(qtiClass).each(function () {
             var $qtiElement = $(this);
             var element = buildElement($qtiElement, options);
 
+            // rendering the table
+            if (qtiClass === _defaultOptions.ns.table) {
+                element = buildTable($qtiElement, options);
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:131]: Please, optimise code. In the case with `table` you call `buildElement` twice

**Review by bziondik — CHANGES_REQUESTED** [review]
1. optimise code
2. add copyright in the beginning of the file
3. update code style to ES6 (var -> const/let), eslint errors

File: src/qtiItem/helper/simpleParser.js (hunk 1 above)
Comment (focaloid-dev) [reply src/qtiItem/helper/simpleParser.js]: code optimized and changed to ES6- for fail safe , i use most of them as let and some of them are in const

File: src/qtiItem/helper/simpleParser.js [hunk 2]
```diff
@@ -102,41 +120,53 @@ function buildTooltip(targetHtml, contentId, contentHtml) {
     };
 }
 
+function buildTable($elt, elt) {
+    elt.body = {
+        body: $elt.html(),
+        elements: {}
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:126]: Here should be parsing elements that inside table, like images

**Review by Silvia-loza — CHANGES_REQUESTED** [review]
Bug: The asset inside the table does not show when editing.
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: src/qtiItem/helper/simpleParser.js [hunk 3]
```diff
@@ -1,35 +1,55 @@
...
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA
+ **/
 import _ from 'lodash';
 import $ from 'jquery';
 import util from 'taoQtiItem/qtiItem/helper/util';
 import Loader from 'taoQtiItem/qtiItem/core/Loader';
 
-var _parsableElements = ['img', 'object', 'printedVariable'];
-var _qtiClassNames = {
+let _parsableElements = ['img', 'object', 'printedVariable'];
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:23]: Use `const` for variables that wouldn't be changed

File: src/qtiItem/helper/simpleParser.js [hunk 4]
```diff
@@ -102,41 +122,83 @@ function buildTooltip(targetHtml, contentId, contentHtml) {
     };
 }
 
+function buildTable($elt, elt, options) {
+    elt.body = {
+        body: '',
+        elements: {}
+    };
+
+    $elt.find(_getElementSelector('math', options.ns.math)).each(function () {
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:131]: You have duplication of code. Please, use `parseContainer`

File: src/qtiItem/helper/simpleParser.js [hunk 5]
```diff
@@ -102,41 +122,83 @@ function buildTooltip(targetHtml, contentId, contentHtml) {
...
         body: '',
         elements: {}
     };
 
-    _.each(_parsableElements, function(qtiClass) {
-        $container.find(qtiClass).each(function() {
-            var $qtiElement = $(this);
-            var element = buildElement($qtiElement, options);
+    _.each(_parsableElements, function (qtiClass) {
+        $container.find(options.ns.table).each(function () {
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:167]: This should be before `_.each`

**Review by bziondik — CHANGES_REQUESTED** [review]
1. use `const` instead of `let`
2. use `parseContainer` for table
and other comment

File: src/qtiItem/helper/simpleParser.js [hunk 6]
```diff
@@ -102,41 +122,68 @@ function buildTooltip(targetHtml, contentId, contentHtml) {
     };
 }
 
+function buildTable($elt, elt, options) {
+    elt.body = {
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:126]: It this part needed?
 `body` and `elements` will be copied later

File: src/qtiItem/helper/simpleParser.js [hunk 7]
```diff
@@ -102,41 +122,68 @@ function buildTooltip(targetHtml, contentId, contentHtml) {
...
-        });
+        element = buildTable($qtiElement, element, options);
+        ret.elements[element.serial] = element;
+        $qtiElement.replaceWith(_placeholder(element));
     });
 
-    $container.find(_getElementSelector('math', options.ns.math)).each(function() {
-        var $qtiElement = $(this);
-        var element = buildMath($qtiElement, options);
+    $container.find(_getElementSelector('img', '')).each(function () {
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:153]: please, use `_.each(_parsableElements, function(qtiClass) {` as it was previously
you skipped `printedVariable` and this is a reason why unit test failed

**Review by bziondik — CHANGES_REQUESTED** [review]
1. use `let` when needed
2. use `_parsableElements`

File: src/qtiItem/helper/simpleParser.js (hunk 6 above)
Comment (focaloid-dev) [reply src/qtiItem/helper/simpleParser.js:126]: yes ,i think it need to be there , unless the assigning of body elements will throw undefined !

File: src/qtiItem/helper/simpleParser.js (hunk 7 above)
Comment (focaloid-dev) [reply src/qtiItem/helper/simpleParser.js:153]: `ret.elements[element.serial] = element;`
some way accidentally deleted this line , this was causing the test case failed

File: src/qtiItem/helper/simpleParser.js (hunk 6 above)
Comment (focaloid-dev) [reply src/qtiItem/helper/simpleParser.js:126]: Object structure needs to be catched before copying , so i think this method is needful

File: src/qtiItem/helper/simpleParser.js [hunk 8]
```diff
@@ -102,41 +122,68 @@ function buildTooltip(targetHtml, contentId, contentHtml) {
...
 
-            ret.elements[element.serial] = element;
-            $qtiElement.replaceWith(_placeholder(element));
-        });
+        element = buildTable($qtiElement, element, options);
+        ret.elements[element.serial] = element;
+        $qtiElement.replaceWith(_placeholder(element));
+    });
+
+    $container.find(_getElementSelector('img', options.ns.image)).each(function () {
```
Comment (bziondik) [inline src/qtiItem/helper/simpleParser.js:153]: Instead handling `img` and `object` separate, just use array `_parsableElements` it already contains `img` and `object`
`_.each(_parsableElements, function (qtiClass) {`

Comment (focaloid-dev) [reply src/qtiItem/helper/simpleParser.js]: Code optimised

**Review by bziondik — APPROVED** [review]
- [ ]  New code is covered by tests (if applicable)
- [x]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x]  New code is respecting code style rules
- [x]  New code is respecting best practices
- [x]  New code is not subject to concurrency issues (if applicable)
- [x]  Feature is working correctly on my local machine (if applicable)
- [x]  Acceptance criteria are respected
- [x]  Pull request title and description are meaningful

**Review by AnastasiaArcadia — APPROVED** [review]
- [ ]  New code is covered by tests (if applicable)
- [x]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x]  New code is respecting code style rules
- [x]  New code is respecting best practices
- [x]  New code is not subject to concurrency issues (if applicable)
- [x]  Feature is working correctly on my local machine (if applicable)
- [x]  Acceptance criteria are respected
- [x]  Pull request title and description are meaningful

**Comment by AntonNeviarouski** [comment]
QA verified on http://tao-test-alshoja.playground.kitchen.it.taocloud.org:46031/

**Pull request merged**